src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -47,21 +50,38 @@ public sealed class RestoreManagerPackage : AsyncPackage
         // won't get disconnected.
         private EnvDTE.BuildEvents _buildEvents;
 
+        // keep a reference to initialize task so that it can complete
+        // even after InitializeAsync() is done.
+        private JoinableTask _initializationTask;
+
         protected override async Task InitializeAsync(
             CancellationToken cancellationToken, 
             IProgress<ServiceProgressData> progress)
         {
             var componentModel = await GetServiceAsync(typeof(SComponentModel)) as IComponentModel;
             componentModel.DefaultCompositionService.SatisfyImportsOnce(this);
 
-            // Accessing DTE without confirming to Main thread because AllowsBackgroundLoading is false which
+            // Start this initialization task but don't await since it will degrade performance.
             // will make sure that this piece is always executed on Main thread.
-            var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
-            _buildEvents = dte.Events.BuildEvents;
-            _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
-
-            UserAgent.SetUserAgentString(
-                new UserAgentStringBuilder().WithVisualStudioSKU(dte.GetFullVsVersionString()));
+            // Besides we don't need to wait here to complete since this will only bind to build event which
+            _initializationTask = ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                try
+                {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
+                    _buildEvents = dte.Events.BuildEvents;
+                    _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
+
+                    UserAgent.SetUserAgentString(
+                        new UserAgentStringBuilder().WithVisualStudioSKU(dte.GetFullVsVersionString()));
+                }
+                catch (Exception ex)
```

Comment:
Yes I meant Package...

`But with this change new detached and delayed initialization task may complete some time later after the package itself is initialized and solution is loaded. So it does increase the risk we'd miss first build(s).`
Completely agree there, if we see that way, then it does increase risk of missing first build event.